zmq.go

```diff
@@ -362,7 +375,31 @@ func Subscribe(addr string, topics []string, timeout time.Duration) (*Conn, erro
 // Receive a message from the publisher. It blocks until a new message is
 // received.
 func (c *Conn) Receive() ([][]byte, error) {
-	return c.readMessage()
+	messages, err := c.readMessage()
+	// If the error is either nil or a non-EOF error, we return it as-is.
+	if err != io.EOF {
+		return messages, err
+	}
+	// We got an EOF, so our socket is disconnected. We attempt to
+	// reconnect. If successful, replace the existing connection with the
+	// new one. Either way, return a timeout error.
+	errTimeout := &net.OpError{
+		Op:     "read",
+		Net:    c.conn.LocalAddr().Network(),
+		Source: c.conn.LocalAddr(),
+		Addr:   c.conn.RemoteAddr(),
+		Err:    &reconnectError{err},
+	}
+	newConn, err := Subscribe(c.conn.RemoteAddr().String(), c.topics,
+		c.timeout)
+	if err != nil {
+		// Prevent CPU overuse by refused reconnection attempts.
+		time.Sleep(c.timeout)
```

Comment:
Perhaps we should use an exponential back off here? For reference the btcsuite rpcclient does something similar: https://github.com/btcsuite/btcd/blob/master/rpcclient/infrastructure.go#L641

Reply:
This is used synchronously, which means each time we back off, we can delay processing of other events in an event loop that include a call to this function. For example, https://github.com/Roasbeef/btcwallet/blob/master/chain/bitcoind.go#L632 is in a loop that also checks for updates to watch lists and rescans. The rpcclient reconnect handler runs as a goroutine.

An option for doing the incremental back off would be to do that in a goroutine, and just sleep/return a timeout error in the synchronous portion.

Reply:
Another option for doing incremental backoff is to add more state to `Conn` to keep track of last attempted reconnection time when disconnected.

Reply:
I see, so even with this patch, we need reconnection logic within the wallet as well? As this just return a timeout error in either case, even if we're able to reconnect successfully.

Reply:
No, the wallet already ignores read timeouts, this code just emulates those: https://github.com/Roasbeef/btcwallet/blob/master/chain/bitcoind.go#L636

Reply:
If we reconnect successfully, the next attempt at reading will return a message if there is one.

Reply:
Ahh, thanks! OK, will test out this patch locally.